src/box3d.rs

```diff
@@ -0,0 +1,939 @@
+use size::TypedSize3D;
+use approxord::{min, max};
+
+use num_traits::NumCast;
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use core::borrow::Borrow;
+use core::cmp::PartialOrd;
+use core::fmt;
+use core::hash::{Hash, Hasher};
+use core::ops::{Add, Div, Mul, Sub, Neg};
+
+
+/// A 3d box optionally tagged with a unit. 
+/// 
+/// a is the upper left front corner position.
+/// b is the lower right back corner position.
+/// 
+/// Axis directions: x axis positive going left to right.
+///                  y axis positive going bottom to top.
+///                  z axis positive going from back to front (out of the page).
+/// The axis directions follows those used by Microsoft:
+/// https://docs.microsoft.com/en-us/dotnet/framework/wpf/graphics-multimedia/3-d-transformations-overview
+#[repr(C)]
+pub struct TypedBox3D<T, U = UnknownUnit> {
+    pub a: TypedPoint3D<T, U>, 
+    pub b: TypedPoint3D<T, U>,
+}
+
+/// The default box 3d type with no unit.
+pub type Box3D<T> = TypedBox3D<T, UnknownUnit>;
+
+#[cfg(feature = "serde")]
+impl<'de, T: Copy + Deserialize<'de>, U> Deserialize<'de> for TypedBox3D<T, U> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let (a, b) = try!(Deserialize::deserialize(deserializer));
+        Ok(TypedBox3D::new(a, b))
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<T: Serialize, U> Serialize for TypedBox3D<T, U> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        (&self.a, &self.b).serialize(serializer)
+    }
+}
+
+impl<T: Hash, U> Hash for TypedBox3D<T, U> {
+    fn hash<H: Hasher>(&self, h: &mut H) {
+        self.a.hash(h);
+        self.b.hash(h);
+    }
+}
+
+impl<T: Copy, U> Copy for TypedBox3D<T, U> {}
+
+impl<T: Copy, U> Clone for TypedBox3D<T, U> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<T: PartialEq, U> PartialEq<TypedBox3D<T, U>> for TypedBox3D<T, U> {
+    fn eq(&self, other: &Self) -> bool {
+        self.a.eq(&other.a) && self.b.eq(&other.b)
+    }
+}
+
+impl<T: Eq, U> Eq for TypedBox3D<T, U> {}
+
+impl<T: fmt::Debug, U> fmt::Debug for TypedBox3D<T, U> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "TypedBox3D({:?}, {:?})", self.a, self.b)
+    }
+}
+
+impl<T: fmt::Display, U> fmt::Display for TypedBox3D<T, U> {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "Box3D({}, {})", self.a, self.b)
+    }
+}
+
+impl<T, U> TypedBox3D<T, U> {
+    /// Constructor.
+    pub fn new(a: TypedPoint3D<T, U>, b: TypedPoint3D<T, U>) -> Self {
+        TypedBox3D {
+            a,
+            b,
+        }
+    }
+}
+
+impl<T, U> TypedBox3D<T, U>
+where
+    T: Copy + Div<T, Output = T> + Neg<Output = T> + Add<T, Output = T> + One,
+{
+    /// Creates a box3d of the given size, so that the centroid is at zero origin.
```

Comment:
Went for option 1, but I'm assuming negative size components are allowed, so we cant just do a direct assignment (eg `box.max = size.to_vector()`). Instead it will have to go through from_points().